File: interactive_replace.py
import re
from pathlib import Path

def replace_interactive(file_path, old_pattern, new_text, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    pattern = re.compile(old_pattern, re.DOTALL)
    matches = pattern.findall(content)
    if not matches:
        raise ValueError(f"Pattern '{old_pattern}' not found in {file_path}")
    print(f"Preview:
Old:
{matches[0]}
New:
{new_text}")
    confirm = input("Apply replacement? (y/n): ").lower()
    if confirm != "y":
        return "Replacement cancelled"
    new_content = pattern.sub(new_text, content)
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(new_content)
    return f"Applied interactive replacement in {file_path}"

File: ts-tools/interactive_replace.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';
import { checkFilePermission } from './check_file_permission';

export function replaceInteractive(
  filePath: string,
  oldPattern: string,
  newText: string,
  config: any,
): Promise<string> {
  return new Promise((resolve, reject) => {
    checkFilePermission(filePath, 'write', config)
      .then(() => {
        const resolvedPath = path.resolve(filePath);
        fs.readFile(resolvedPath, 'utf-8', (err, data) => {
          if (err) {
            reject(err);
          } else {
            const regex = new RegExp(oldPattern, 's');
            const match = data.match(regex);
            if (!match) {
              reject(
                new Error(`Pattern '${oldPattern}' not found in ${filePath}`),
              );
              return;
            }
            console.log(`Preview:\nOld:\n${match[0]}\nNew:\n${newText}`);
            const rl = readline.createInterface({
              input: process.stdin,
              output: process.stdout,
            });
            rl.question('Apply replacement? (y/n): ', (answer) => {
              if (answer.toLowerCase() !== 'y') {
                resolve('Replacement cancelled');
              } else {
                const newContent = data.replace(regex, newText);
                fs.writeFile(resolvedPath, newContent, 'utf-8', (err) => {
                  if (err) {
                    reject(err);
                  } else {
                    resolve(`Applied interactive replacement in ${filePath}`);
                  }
                });
              }
              rl.close();
            });
          }
        });
      })
      .catch(reject);
  });
}
